apps/remix-ide/src/app/components/top-bar.tsx

```diff
@@ -0,0 +1,118 @@
+/* eslint-disable @nrwl/nx/enforce-module-boundaries */
+import React from 'react'
+import { TopbarProvider } from '@remix-ui/top-bar'
+import packageJson from '../../../../../package.json'
+import { EventEmitter } from 'events'
+import { Plugin } from '@remixproject/engine'
+import { PluginViewWrapper } from '@remix-ui/helper'
+import { AppAction } from 'libs/remix-ui/app/src/lib/remix-app/actions/app'
+import FilePanel from '../panels/file-panel'
+import { WorkspaceMetadata } from 'libs/remix-ui/workspace/src/lib/types'
+import { gitUIPanels } from '@remix-ui/git'
+import { HOME_TAB_NEW_UPDATES } from 'libs/remix-ui/home-tab/src/lib/components/constant'
+import axios from 'axios'
+import { UpdateInfo } from 'libs/remix-ui/home-tab/src/lib/components/types/carouselTypes'
+import { GitPlugin } from '../plugins/git'
+
+const TopBarProfile = {
+  name: 'topbar',
+  displayName: 'Top Bar',
+  description: '',
+  version: packageJson.version,
+  icon: '',
+  methods: [],
+  events: []
+}
+
+export class Topbar extends Plugin {
+  dispatch: React.Dispatch<any> = () => { }
+  appStateDispatch: React.Dispatch<AppAction> = () => { }
+  htmlElement: HTMLDivElement
+  events: EventEmitter
+  topbarExpandPath: string
+  filePanel: FilePanel
+  git: GitPlugin
+  workspaces: WorkspaceMetadata[]
+  currentWorkspaceMetadata: WorkspaceMetadata
+
+  constructor(filePanel: FilePanel, git: GitPlugin) {
+    super(TopBarProfile)
+    this.filePanel = filePanel
+    this.git = git
+    this.workspaces = []
+    this.currentWorkspaceMetadata = null
+  }
+
+  onActivation(): void {
+    this.renderComponent()
+  }
+
+  onDeactivation(): void {
+
+  }
+
+  async getWorkspaces() {
+    while (this.workspaces.length === 0) {
```

Comment:
not sure if we could use Event listing instead of this while loop

Reply:
Done